Прогресс 21.06

TheGodseeker opened the pull request.

TheGodseeker committed:
Начало работы над контроллером

TheGodseeker committed:
Сделана связь сервера и клиента

- Добавлен контроллер (файл Controller.cs)
- Создано отображение полученных данных от сервера о всех задачах на клиенте

TheGodseeker committed:
Сделано отображение выполненных и невыполненных задач

- Наладил связь между клиентом и сервером для отображения выполненных и невыполненных задач
- Почистил лишние комментарии

TheGodseeker committed:
Создание механизма создания новых задач

- Теперь страница для создания новых задач отправляет данные на сервер, чтобы новая задача появилась в БД

TheGodseeker committed:
Добавление отображения промежутков времени у задач

TheGodseeker committed:
Добавление возможности изменения статуса задачи

- Добавлена функция на нажатие кнопки "Закончить" возле задачи. Теперь можно изменить статус задачи на "Выполнено"
- Добавлены секунды в отображение времени на задачу

TheGodseeker committed:
Обновил БД

- Добавил столбец isActive в таблицу timeGaps

TheGodseeker committed:
Добавление создания промежутка работы

- Сделал возможность создать временной промежуток (пока только его начало)

AndreyGurev commented on timetrackerUI/styles/styles.css:
один файл со стилями - идея плохая. Представь, что твое приложение будет в 100 раз больше. Такой файл невозможно будет читать. 
Лучше делать файл стилей отдельно для каждой компоненты. 
Общие стили для всех лучше так и называть Shared или Common styles.

AndreyGurev commented on timetracker/Controller.cs:
Контроллеров в проекте обычно очень много, каждый отвечает за свою область. 
Этот твой, за задачи, поэтому так и назови его - TaskController

AndreyGurev replied:
Тебе нужно разнести логику работы с тасками и логику в контроллере. Писанины будет больше, но поддерживаемость  кода улучшиться. Начни хотя бы с репозитория. 
https://habr.com/ru/articles/248505/

TheWhitewind replied:
Разделил на два отдельных контроллера: для задач и для промежутков времени.

TheWhitewind replied:
Сделано

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+   {
+        public class TimeTracker : Controller
+        {
+            
```
старайся не допускать таких помарок. Ниже у тебя 4 строки пустые. Это может усложнять чтение

TheWhitewind replied:
Поправил

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            {
+                using (var context = new DBConnector()){
+                    List<Task> tasks = new List<Task>();
+                    foreach(var task in context.tasks.ToList()){
```
Используй фильтры при запросе. Незачем доставать ВСЕ таски и фильтровать их в приложении.
Используй IQueryable. В данном случае подойдет Where

TheWhitewind replied:
Сделано

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+                        if (task.isDone)
+                            tasks.Add(task);
+                    }
+                    string result = JsonSerializer.Serialize(tasks);
```
не указывай явно типы. Хотя некоторые и рекомендуют так делать, все таки это не очень хорошая практика в современном мире - код становится громоздким, ухудшается читабельность. 
тем более, выше ты используешь var. Либо var либо точные типы

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            {
+                using (var context = new DBConnector()){
+                    List<Task> tasks = new List<Task>();
+                    foreach(var task in context.tasks.ToList()){
```
тоже самое, используй фильтры запросов

TheWhitewind replied:
Сделано

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            }
+
+            [HttpPost]
+            public void AddTask(string _name, string _desc = null, long exptTime = 0)
```
подчерки "_" используются для приватных полей и свойств, не для параметров

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            [HttpPut]
+            public void UpdateEndTimeGap(long gapId, bool isAct = true)
+            {
+                using (var context = new DBConnector()){
```
using можно использовать без {}, так ты избегаешь вложенности.
Используй var using

TheWhitewind replied:
Сделано

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            [HttpGet]
+            public IActionResult GetAllTasks()
+            {
+                using (var context = new DBConnector()){
```
Используй DI. https://metanit.com/sharp/aspnet5/6.1.php
Так тебе не придется каждый раз объявлять DBConnector в методах, ты просто передаешь его в конструктор

TheWhitewind replied:
Сделано

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+            }
+
+            /*
+            TO-DO:
```
Молодец! Такие тудухи хорошая практика - позволяет не забыть важные вещи тебе и остальным участникам проекта

AndreyGurev commented on timetracker/Controller.cs:
```diff
@@ -0,0 +1,187 @@
+                using (var context = new DBConnector()){
+                    var curGap = context.timeGaps.FirstOrDefault(item => item.id == gapId);
+                
+                    if (curGap != null)
```
а если curGap  == null? какую ошибку получить клиент?
Лучше обрабатывать такие частные случаи внятными ошибками.

TheWhitewind replied:
Сделал. Теперь функция возвращает IActionResult: пустой "Ok", если все прошло успешно, и "NoContent", если что-то пошло не так

TheGodseeker committed:
Разделил контроллер

Разделил контроллер на два - отдельно для задач и промежутков

TheGodseeker committed:
Удаление лишнего кода из TasksCotroller

TheGodseeker committed:
Перевод контекста контроллеров в конструкторы

TheGodseeker committed:
Перепись поиска нужных данных

Переписал функции GetDoneTasks, GetUndoneTasks и GetTimeGaps под использование  IQueryable.

TheGodseeker committed:
Реализация DI

Переписал код под использование DI (интерфейсы и БД)

TheGodseeker committed:
Фикс БД

- Теперь клиент получает всю нужную информацию от сервера
- Поправил нейминг репозиториев

TheGodseeker committed:
Установка промежутков времени

- Теперь можно установить начало и конец промежутков работы
- Вычисляется проведенное время работы

TheGodseeker committed:
Чистка кода от лишних комментариев
